Userland/useradd.cpp

```diff
@@ -0,0 +1,155 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "home_dir", "home directory for the new user");
+    args_parser.add_arg("u", "uid", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "group", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "shell", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "name", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return 0;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s\n", args.get("u").characters());
+            return 3;
+        }
+
+        if (getpwuid(uid)) {
+            fprintf(stderr, "uid %u already exists!\n", uid);
+            return 4;
+        }
+
+    } else {
+        for (uid = BASE_UID; getpwuid(uid); uid++) {
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s\n", args.get("g").characters());
+            return 1;
+        }
+    }
+
+    String shell = args.is_present("s") ? args.get("s") : String(DEFAULT_SHELL);
+    String home = args.is_present("d") ? args.get("d") : String("/");
+    String gecos = args.is_present("n") ? args.get("n") : String("Anonymous");
+
+    // Let's run a quick sanity check on username
+    if(strpbrk(username, "\\/123456789!@#$%^&*()~-_+=`:\n"))
+    {
+        fprintf(stderr, "invalid character in username %s\n", username);
+        return 1;
+    }
+
+    FILE* pwfile = fopen("/etc/passwd", "a");
+    if (!pwfile) {
+        perror("failed to open /etc/passwd");
+        return 1;
+    }
+
+    if (args.is_present("m")) {
+        StringBuilder builder;
+        builder.append("/home/");
+        builder.append(username);
+
+        home = builder.to_string(); // This overwrites the -d argument
+        if (mkdir(home.characters(), 0700)) {
+            builder.append("failed to create directory ");
```

Comment:
I wouldn't rely on the fact that `to_string()` clears out the `StringBuilder`. Either create a fresh one or call `.clear()` explicitly.

Reply:
```
String StringBuilder::to_string()
{
    auto string = String((const char*)m_buffer.data(), m_length);
    clear();
    return string;
}
```

Reply:
Besides, why not simply use `String::format()`? `StringBuilder` is the right API when you concatenating strings in a loop or something.

Reply:
> Is this not explicit????

It's explicit in the implementation of `StringBuilder::to_string()`, not in your code. I'd consider it an implementation detail that you should be aware of, but not rely on, and that could change someday.

Actually, why does it even do that, @awesomekling? If it copies the buffer anyway instead of moving it to the new string, why not leave the buffer in place after this call?

Reply:
Yeah we can just leave it in place! It's just a leftover behavior from how `StringBuilder::to_string()` at one point would transfer ownership of the internal buffer to the caller.

Reply:
> Besides, why not simply use String::format()

I didn't know that was a function. I'll definitely be using that